CorsixTH/Lua/HeatingController.lua

```diff
@@ -0,0 +1,182 @@
+end
+
+function HeatingController:hotWarning(isPlayerHospital)
+  local announcements = {
+    "sorry003.wav", "sorry004.wav",
+  }
+  if announcements and isPlayerHospital then
+    self.world.ui:playAnnouncement(announcements[math.random(1, #announcements)])
+  end
+end
+
+-- Called when the hospitals's boiler has broken down.
+-- It will remain broken for a certain period of time.
+function HeatingController:boilerBreakdown(isPlayerHospital)
+  self.curr_setting = self.radiator_heat
+  self.radiator_heat = math.random(0, 1)
+  self.boiler_countdown = math.random(10, 30)
+
+  self.heating_broke = true
+
+  -- Only show the message when relevant to the local player's hospital.
+  if isPlayerHospital then
+    if self.radiator_heat == 0 then
+      self.world.ui.adviser:say(_A.boiler_issue.minimum_heat)
+      self:coldWarning(isPlayerHospital)
+    else
+      self.world.ui.adviser:say(_A.boiler_issue.maximum_heat)
+      self:hotWarning(isPlayerHospital)
+    end
+  end
+end
+
+-- When the boiler has been repaired this function is called.
+function HeatingController:boilerFixed(isPlayerHospital)
+  self.radiator_heat = self.curr_setting
+  self.heating_broke = false
+  self:resetBoilerBreakdownCounter()
+  if isPlayerHospital then
+    self.world.ui.adviser:say(_A.boiler_issue.resolved)
+  end
+end
+
+function HeatingController:onEndDay(nbHandyman, isPlayerHospital)
+  --variables for heating
+  local radiators = self.world.object_counts.radiator
+
+  -- Countdown for boiler breakdowns
+  if self.heating_broke then
+    if 5 * nbHandyman >= radiators and self.boiler_countdown > 3 then
+      self.boiler_countdown = self.boiler_countdown - 3
+    elseif 8 * nbHandyman >= radiators and self.boiler_countdown > 2 then
+      self.boiler_countdown = self.boiler_countdown - 2
+    else
+      self.boiler_countdown = self.boiler_countdown - 1
+    end
+    if self.boiler_countdown == 0 then
+      self:boilerFixed(isPlayerHospital)
+    end
+  end
+
+  -- Is the boiler working today?
+  if self.days_until_boiler_breakdown <= 0 then
+    self:resetBoilerBreakdownCounter()
+    --33% chance nothing happen 66% Boiler breakdown
+    --Todo: 50% chance boiler breakdown, 25% chance vomit wave, 25% change nothing happen
+    if math.random(1,3) ~= 1 then
+      --check if boiler can break
+      if not self.heating_broke and self.boiler_can_break and radiators > 0 and 8 * nbHandyman < radiators then
+        self:boilerBreakdown(isPlayerHospital)
+      end
+    end
+  else
+    self.days_until_boiler_breakdown = self.days_until_boiler_breakdown - 1
+  end
+
+  -- Calculate heating cost daily.  Divide the monthly cost by the number of days in that month
+  local heating_costs = (((self.radiator_heat * 10) * radiators) * 7.50) / self.world:date():lastDayOfMonth()
+  self.acc_heating = self.acc_heating + heating_costs
+end
+
+function HeatingController:getHeatingCostsForActualMonth()
+  local monthlyCosts = math.round(self.acc_heating)
+  self.acc_heating = 0
```

Comment:
Getter should not have side effects. Reset acc_heating in setter or onEndMonth method.